src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstancesLoaderImpl.scala

```diff
@@ -4,26 +4,59 @@ package core.task.tracker.impl
 import akka.stream.Materializer
 import akka.stream.scaladsl.Sink
 import com.typesafe.scalalogging.StrictLogging
-import mesosphere.marathon.core.task.tracker.InstanceTracker
-import mesosphere.marathon.storage.repository.InstanceRepository
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.task.tracker.{InstanceTracker, InstanceTrackerConfig}
+import mesosphere.marathon.state.RunSpec
+import mesosphere.marathon.storage.repository.{GroupRepository, InstanceRepository}
 
+import scala.async.Async.{async, await}
 import scala.concurrent.Future
 
 /**
   * Loads all task data into an [[InstanceTracker.InstancesBySpec]] from an [[InstanceRepository]].
   */
-private[tracker] class InstancesLoaderImpl(repo: InstanceRepository)(implicit val mat: Materializer)
+private[tracker] class InstancesLoaderImpl(
+    repo: InstanceRepository,
+    groupRepository: GroupRepository,
+    config: InstanceTrackerConfig)(implicit val mat: Materializer)
   extends InstancesLoader with StrictLogging {
   import scala.concurrent.ExecutionContext.Implicits.global
 
-  override def load(): Future[InstanceTracker.InstancesBySpec] = {
-    for {
-      names <- repo.ids().runWith(Sink.seq)
-      _ = logger.info(s"About to load ${names.size} instances")
-      instances <- Future.sequence(names.map(repo.get)).map(_.flatten)
-    } yield {
-      logger.info(s"Loaded ${instances.size} instances")
-      InstanceTracker.InstancesBySpec.forInstances(instances)
-    }
+  override def load(): Future[InstanceTracker.InstancesBySpec] = async {
+    val instances = repo.ids().mapAsync(parallelism = config.internalInstanceTrackerNumParallelLoads())(repo.get).mapConcat(_.toList)
+
+    // Join instances with app or pod.
+    val coreInstances = await(
+      instances.mapAsync(parallelism = config.internalInstanceTrackerNumParallelLoads()) { stateInstance =>
+        async {
+          val runSpecId = stateInstance.instanceId.runSpecId
+          val runSpecVersion = stateInstance.runSpecVersion.toOffsetDateTime
+          await(groupRepository.runSpecVersion(runSpecId, runSpecVersion)) match {
+            case Some(runSpec) => (stateInstance, Some(runSpec))
+            case None =>
+              logger.warn(s"No run spec $runSpecId with version ${stateInstance.runSpecVersion} was found for instance ${stateInstance.instanceId}. Trying latest.")
+              await(groupRepository.latestRunSpec(runSpecId)) match {
+                case Some(runSpec) => (stateInstance, Some(runSpec))
+                case None => await(expungeOrphanedInstance(stateInstance))
```

Comment:
Nothing is listening at this point. The instance tracker is one of the first things being loaded. Most other entities assume a running instance tracker. I understand that we could handle this somewhere else but then again, what part in our code base should deal with orphaned instances and why? Ideally this never happens to begin with.

On a meta level I do not like our approach with events. However, that is something we'll discuss for our scheduler interface, I guess.

Reply:
for me it's like garbage collection, so I would have a separate entity (actor or whatever) doing that cleaning. That would be triggered e.g. by event. You don't call java garbage collection in java directly as well...

Reply:
> You don't call java garbage collection in java directly as well...

unless you do `System.gc()` ;)

Like Karsten said - nothing is listening at that point, and the one responsible for expunging them is the instance tracker itself so ¯\_(ツ)_/¯

Reply:
I get your point @alenkacz. However,
* We would have to load them first to garbage collect. That would require some run spec or `Option.empty[RunSpec]` or something.
* It would complexity we might not want.

Reply:
Hm, I could add a prune step and then load with an adapter repo that has access to run specs and instances. Would that be ok?

Reply:
btw. this is not a blocker for me :) I am just expressing my "gold state" wish for that piece of code. I don't mean we have to do this in this PR... I would be fine with merging this but had a JIRA/googledoc or whatever with some plan on how to improve that :)